package eu.isas.peptideshaker.gui.tabpanels;

import eu.isas.peptideshaker.gui.PeptideShakerGUI;
import eu.isas.peptideshaker.utils.BareBonesBrowserLaunch;

/**
 * This tab contains the basic protein annotation and links to other 
 * protein resources.
 * 
 * @author Harald Barsnes
 */
public class AnnotationPanel extends javax.swing.JPanel {

    private PeptideShakerGUI peptideShakerGUI;

    /** 
     * Creates a new AnnotationPanel.
     * 
     * @param peptideShakerGUI
     * @param coverage  
     */
    public AnnotationPanel(PeptideShakerGUI peptideShakerGUI) {
        initComponents();


        this.peptideShakerGUI = peptideShakerGUI;
    }

    /** This method is called from within the constructor to
     * initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is
     * always regenerated by the Form Editor.
     */
    @SuppressWarnings("unchecked")
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    private void initComponents() {

        annotationLinksJPanel = new javax.swing.JPanel();
        jLabel1 = new javax.swing.JLabel();
        accessionNumberJTextField = new javax.swing.JTextField();
        jLabel3 = new javax.swing.JLabel();
        accessionNumberJTextField1 = new javax.swing.JTextField();
        jLabel4 = new javax.swing.JLabel();
        accessionNumberJTextField2 = new javax.swing.JTextField();
        jLabel8 = new javax.swing.JLabel();
        accessionNumberJTextField5 = new javax.swing.JTextField();
        jScrollPane1 = new javax.swing.JScrollPane();
        jTextArea1 = new javax.swing.JTextArea();
        annotationLinksJPanel1 = new javax.swing.JPanel();
        jLabel5 = new javax.swing.JLabel();
        loadUniProtJButton = new javax.swing.JButton();
        annotationLinksJPanel2 = new javax.swing.JPanel();
        jLabel7 = new javax.swing.JLabel();
        loadDastyJButton = new javax.swing.JButton();
        annotationLinksJPanel3 = new javax.swing.JPanel();
        jLabel9 = new javax.swing.JLabel();
        loadReactomeJButton = new javax.swing.JButton();
        annotationLinksJPanel4 = new javax.swing.JPanel();
        jLabel10 = new javax.swing.JLabel();
        loadStringJButton = new javax.swing.JButton();
        annotationLinksJPanel5 = new javax.swing.JPanel();
        jLabel11 = new javax.swing.JLabel();
        loadDavidJButton = new javax.swing.JButton();
        annotationLinksJPanel6 = new javax.swing.JPanel();
        jLabel12 = new javax.swing.JLabel();
        loadDavidJButton1 = new javax.swing.JButton();
        annotationLinksJPanel7 = new javax.swing.JPanel();
        jLabel13 = new javax.swing.JLabel();
        loadDavidJButton2 = new javax.swing.JButton();
        annotationLinksJPanel8 = new javax.swing.JPanel();
        jLabel14 = new javax.swing.JLabel();
        loadDavidJButton3 = new javax.swing.JButton();
        jPanel1 = new javax.swing.JPanel();
        jLabel2 = new javax.swing.JLabel();
        jLabel15 = new javax.swing.JLabel();
        jLabel17 = new javax.swing.JLabel();

        setBackground(new java.awt.Color(255, 255, 255));

        annotationLinksJPanel.setBorder(javax.swing.BorderFactory.createTitledBorder("Basic Protein Annotation"));
        annotationLinksJPanel.setOpaque(false);

        jLabel1.setText("Accession Number:");

        accessionNumberJTextField.setEditable(false);
        accessionNumberJTextField.setHorizontalAlignment(javax.swing.JTextField.CENTER);
        accessionNumberJTextField.setText("P02768");

        jLabel3.setText("Protein Name:");

        accessionNumberJTextField1.setEditable(false);
        accessionNumberJTextField1.setHorizontalAlignment(javax.swing.JTextField.CENTER);
        accessionNumberJTextField1.setText("(not yet implemented)");

        jLabel4.setText("Gene Name(s):");

        accessionNumberJTextField2.setEditable(false);
        accessionNumberJTextField2.setHorizontalAlignment(javax.swing.JTextField.CENTER);
        accessionNumberJTextField2.setText("(not yet implemented)");

        jLabel8.setText("Taxonomy:");

        accessionNumberJTextField5.setEditable(false);
        accessionNumberJTextField5.setHorizontalAlignment(javax.swing.JTextField.CENTER);
        accessionNumberJTextField5.setText("(not yet implemented)");

        jTextArea1.setColumns(20);
        jTextArea1.setEditable(false);
        jTextArea1.setLineWrap(true);
        jTextArea1.setRows(5);
        jTextArea1.setText("(not yet implemented)");
        jTextArea1.setWrapStyleWord(true);
        jScrollPane1.setViewportView(jTextArea1);

        javax.swing.GroupLayout annotationLinksJPanelLayout = new javax.swing.GroupLayout(annotationLinksJPanel);
        annotationLinksJPanel.setLayout(annotationLinksJPanelLayout);
        annotationLinksJPanelLayout.setHorizontalGroup(
            annotationLinksJPanelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanelLayout.createSequentialGroup()
                .addContainerGap()
                .addGroup(annotationLinksJPanelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(jLabel1)
                    .addComponent(jLabel3)
                    .addComponent(jLabel4)
                    .addComponent(jLabel8))
                .addGap(18, 18, 18)
                .addGroup(annotationLinksJPanelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(accessionNumberJTextField5, javax.swing.GroupLayout.DEFAULT_SIZE, 183, Short.MAX_VALUE)
                    .addComponent(accessionNumberJTextField2, javax.swing.GroupLayout.DEFAULT_SIZE, 183, Short.MAX_VALUE)
                    .addComponent(accessionNumberJTextField1, javax.swing.GroupLayout.DEFAULT_SIZE, 183, Short.MAX_VALUE)
                    .addComponent(accessionNumberJTextField, javax.swing.GroupLayout.DEFAULT_SIZE, 183, Short.MAX_VALUE))
                .addGap(31, 31, 31)
                .addComponent(jScrollPane1, javax.swing.GroupLayout.DEFAULT_SIZE, 226, Short.MAX_VALUE)
                .addContainerGap())
        );
        annotationLinksJPanelLayout.setVerticalGroup(
            annotationLinksJPanelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanelLayout.createSequentialGroup()
                .addContainerGap()
                .addGroup(annotationLinksJPanelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(jScrollPane1, javax.swing.GroupLayout.DEFAULT_SIZE, 114, Short.MAX_VALUE)
                    .addGroup(annotationLinksJPanelLayout.createSequentialGroup()
                        .addGroup(annotationLinksJPanelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                            .addComponent(jLabel1)
                            .addComponent(accessionNumberJTextField, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addGroup(annotationLinksJPanelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                            .addComponent(jLabel3)
                            .addComponent(accessionNumberJTextField1, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addGroup(annotationLinksJPanelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                            .addComponent(jLabel4)
                            .addComponent(accessionNumberJTextField2, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addGroup(annotationLinksJPanelLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                            .addComponent(accessionNumberJTextField5, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                            .addComponent(jLabel8))))
                .addContainerGap())
        );

        annotationLinksJPanel1.setBorder(javax.swing.BorderFactory.createTitledBorder("UniProt - protein knowledgebase"));
        annotationLinksJPanel1.setOpaque(false);

        jLabel5.setText("High-quality protein sequence and functional information.");

        loadUniProtJButton.setText("Search UniProt");
        loadUniProtJButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                loadUniProtJButtonActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout annotationLinksJPanel1Layout = new javax.swing.GroupLayout(annotationLinksJPanel1);
        annotationLinksJPanel1.setLayout(annotationLinksJPanel1Layout);
        annotationLinksJPanel1Layout.setHorizontalGroup(
            annotationLinksJPanel1Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel1Layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(loadUniProtJButton, javax.swing.GroupLayout.PREFERRED_SIZE, 130, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(18, 18, 18)
                .addComponent(jLabel5)
                .addContainerGap(136, Short.MAX_VALUE))
        );
        annotationLinksJPanel1Layout.setVerticalGroup(
            annotationLinksJPanel1Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel1Layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(annotationLinksJPanel1Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(loadUniProtJButton)
                    .addComponent(jLabel5))
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );

        annotationLinksJPanel2.setBorder(javax.swing.BorderFactory.createTitledBorder("Dasty 3 - protein sequence features"));
        annotationLinksJPanel2.setOpaque(false);

        jLabel7.setText("Web client for visualizing protein sequence feature information.");

        loadDastyJButton.setText("Search Dasty 3");
        loadDastyJButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                loadDastyJButtonActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout annotationLinksJPanel2Layout = new javax.swing.GroupLayout(annotationLinksJPanel2);
        annotationLinksJPanel2.setLayout(annotationLinksJPanel2Layout);
        annotationLinksJPanel2Layout.setHorizontalGroup(
            annotationLinksJPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel2Layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(loadDastyJButton, javax.swing.GroupLayout.PREFERRED_SIZE, 130, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(18, 18, 18)
                .addComponent(jLabel7)
                .addContainerGap(107, Short.MAX_VALUE))
        );
        annotationLinksJPanel2Layout.setVerticalGroup(
            annotationLinksJPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel2Layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(annotationLinksJPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(jLabel7)
                    .addComponent(loadDastyJButton))
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );

        annotationLinksJPanel3.setBorder(javax.swing.BorderFactory.createTitledBorder("Reactome - pathway database"));
        annotationLinksJPanel3.setOpaque(false);

        jLabel9.setText("Manually curated and peer-reviewed pathway database.");

        loadReactomeJButton.setText("Search Reactome");
        loadReactomeJButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                loadReactomeJButtonActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout annotationLinksJPanel3Layout = new javax.swing.GroupLayout(annotationLinksJPanel3);
        annotationLinksJPanel3.setLayout(annotationLinksJPanel3Layout);
        annotationLinksJPanel3Layout.setHorizontalGroup(
            annotationLinksJPanel3Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel3Layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(loadReactomeJButton, javax.swing.GroupLayout.PREFERRED_SIZE, 130, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(18, 18, 18)
                .addComponent(jLabel9)
                .addContainerGap(138, Short.MAX_VALUE))
        );
        annotationLinksJPanel3Layout.setVerticalGroup(
            annotationLinksJPanel3Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel3Layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(annotationLinksJPanel3Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(jLabel9)
                    .addComponent(loadReactomeJButton))
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );

        annotationLinksJPanel4.setBorder(javax.swing.BorderFactory.createTitledBorder("STRING - protein interaction"));
        annotationLinksJPanel4.setOpaque(false);

        jLabel10.setText("Known and Predicted Protein-Protein Interactions.");

        loadStringJButton.setText("Search STRING");
        loadStringJButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                loadStringJButtonActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout annotationLinksJPanel4Layout = new javax.swing.GroupLayout(annotationLinksJPanel4);
        annotationLinksJPanel4.setLayout(annotationLinksJPanel4Layout);
        annotationLinksJPanel4Layout.setHorizontalGroup(
            annotationLinksJPanel4Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel4Layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(loadStringJButton, javax.swing.GroupLayout.PREFERRED_SIZE, 130, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(18, 18, 18)
                .addComponent(jLabel10)
                .addContainerGap(170, Short.MAX_VALUE))
        );
        annotationLinksJPanel4Layout.setVerticalGroup(
            annotationLinksJPanel4Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel4Layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(annotationLinksJPanel4Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(jLabel10)
                    .addComponent(loadStringJButton))
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );

        annotationLinksJPanel5.setBorder(javax.swing.BorderFactory.createTitledBorder("DAVID - functional annotation"));
        annotationLinksJPanel5.setOpaque(false);

        jLabel11.setText("Database for Annotation, Visualization and Integrated Discovery.");

        loadDavidJButton.setText("Search DAVID");
        loadDavidJButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                loadDavidJButtonActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout annotationLinksJPanel5Layout = new javax.swing.GroupLayout(annotationLinksJPanel5);
        annotationLinksJPanel5.setLayout(annotationLinksJPanel5Layout);
        annotationLinksJPanel5Layout.setHorizontalGroup(
            annotationLinksJPanel5Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel5Layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(loadDavidJButton, javax.swing.GroupLayout.PREFERRED_SIZE, 130, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(18, 18, 18)
                .addComponent(jLabel11)
                .addContainerGap(97, Short.MAX_VALUE))
        );
        annotationLinksJPanel5Layout.setVerticalGroup(
            annotationLinksJPanel5Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel5Layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(annotationLinksJPanel5Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(jLabel11)
                    .addComponent(loadDavidJButton))
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );

        annotationLinksJPanel6.setBorder(javax.swing.BorderFactory.createTitledBorder("IntAct - protein interaction"));
        annotationLinksJPanel6.setOpaque(false);

        jLabel12.setText("Analysis tools for protein interaction data.");

        loadDavidJButton1.setText("Search IntAct");
        loadDavidJButton1.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                loadDavidJButton1ActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout annotationLinksJPanel6Layout = new javax.swing.GroupLayout(annotationLinksJPanel6);
        annotationLinksJPanel6.setLayout(annotationLinksJPanel6Layout);
        annotationLinksJPanel6Layout.setHorizontalGroup(
            annotationLinksJPanel6Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel6Layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(loadDavidJButton1, javax.swing.GroupLayout.PREFERRED_SIZE, 130, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(18, 18, 18)
                .addComponent(jLabel12)
                .addContainerGap(209, Short.MAX_VALUE))
        );
        annotationLinksJPanel6Layout.setVerticalGroup(
            annotationLinksJPanel6Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel6Layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(annotationLinksJPanel6Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(jLabel12)
                    .addComponent(loadDavidJButton1))
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );

        annotationLinksJPanel7.setBorder(javax.swing.BorderFactory.createTitledBorder("QuickGO - gene ontology terms and annotations"));
        annotationLinksJPanel7.setOpaque(false);

        jLabel13.setText("Web-based browser for Gene Ontology terms.");

        loadDavidJButton2.setText("Search QuickGO");
        loadDavidJButton2.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                loadDavidJButton2ActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout annotationLinksJPanel7Layout = new javax.swing.GroupLayout(annotationLinksJPanel7);
        annotationLinksJPanel7.setLayout(annotationLinksJPanel7Layout);
        annotationLinksJPanel7Layout.setHorizontalGroup(
            annotationLinksJPanel7Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel7Layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(loadDavidJButton2, javax.swing.GroupLayout.PREFERRED_SIZE, 130, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(18, 18, 18)
                .addComponent(jLabel13)
                .addContainerGap(188, Short.MAX_VALUE))
        );
        annotationLinksJPanel7Layout.setVerticalGroup(
            annotationLinksJPanel7Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel7Layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(annotationLinksJPanel7Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(jLabel13)
                    .addComponent(loadDavidJButton2))
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );

        annotationLinksJPanel8.setBorder(javax.swing.BorderFactory.createTitledBorder("InterPro - predictive protein signatures"));
        annotationLinksJPanel8.setOpaque(false);

        jLabel14.setText("Integrated database of predictive protein signatures.");

        loadDavidJButton3.setText("Search InterPro");
        loadDavidJButton3.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                loadDavidJButton3ActionPerformed(evt);
            }
        });

        javax.swing.GroupLayout annotationLinksJPanel8Layout = new javax.swing.GroupLayout(annotationLinksJPanel8);
        annotationLinksJPanel8.setLayout(annotationLinksJPanel8Layout);
        annotationLinksJPanel8Layout.setHorizontalGroup(
            annotationLinksJPanel8Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel8Layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(loadDavidJButton3, javax.swing.GroupLayout.PREFERRED_SIZE, 130, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(18, 18, 18)
                .addComponent(jLabel14)
                .addContainerGap(154, Short.MAX_VALUE))
        );
        annotationLinksJPanel8Layout.setVerticalGroup(
            annotationLinksJPanel8Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(annotationLinksJPanel8Layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(annotationLinksJPanel8Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(jLabel14)
                    .addComponent(loadDavidJButton3))
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );

        jPanel1.setBorder(javax.swing.BorderFactory.createTitledBorder("Annotation Help"));
        jPanel1.setOpaque(false);

        jLabel2.setFont(jLabel2.getFont().deriveFont((jLabel2.getFont().getStyle() | java.awt.Font.ITALIC)));
        jLabel2.setText("Get further protein annotations by querying the various resources below.");

        jLabel15.setFont(jLabel15.getFont().deriveFont((jLabel15.getFont().getStyle() | java.awt.Font.ITALIC)));
        jLabel15.setText("(more details to come)");

        jLabel17.setFont(jLabel17.getFont().deriveFont((jLabel17.getFont().getStyle() | java.awt.Font.ITALIC)));
        jLabel17.setForeground(new java.awt.Color(255, 0, 0));
        jLabel17.setText("This tab is still in development...");

        javax.swing.GroupLayout jPanel1Layout = new javax.swing.GroupLayout(jPanel1);
        jPanel1.setLayout(jPanel1Layout);
        jPanel1Layout.setHorizontalGroup(
            jPanel1Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(jPanel1Layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(jPanel1Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(jLabel2)
                    .addComponent(jLabel15)
                    .addComponent(jLabel17))
                .addContainerGap(206, Short.MAX_VALUE))
        );
        jPanel1Layout.setVerticalGroup(
            jPanel1Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(jPanel1Layout.createSequentialGroup()
                .addContainerGap()
                .addComponent(jLabel2)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addComponent(jLabel15)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED, 66, Short.MAX_VALUE)
                .addComponent(jLabel17)
                .addContainerGap())
        );

        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(this);
        this.setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                    .addComponent(annotationLinksJPanel7, javax.swing.GroupLayout.Alignment.LEADING, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(annotationLinksJPanel4, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(annotationLinksJPanel3, javax.swing.GroupLayout.Alignment.LEADING, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(annotationLinksJPanel1, javax.swing.GroupLayout.Alignment.LEADING, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(annotationLinksJPanel, javax.swing.GroupLayout.Alignment.LEADING, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                    .addComponent(jPanel1, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(annotationLinksJPanel6, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(annotationLinksJPanel5, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(annotationLinksJPanel2, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(annotationLinksJPanel8, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
                .addContainerGap())
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING, false)
                    .addComponent(jPanel1, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addComponent(annotationLinksJPanel, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(annotationLinksJPanel1, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(annotationLinksJPanel2, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                    .addComponent(annotationLinksJPanel5, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(annotationLinksJPanel3, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(annotationLinksJPanel4, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(annotationLinksJPanel6, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(annotationLinksJPanel7, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(annotationLinksJPanel8, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );
    }// </editor-fold>//GEN-END:initComponents

    private void loadDastyJButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_loadDastyJButtonActionPerformed
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.WAIT_CURSOR));
        BareBonesBrowserLaunch.openURL("http://www.ebi.ac.uk/dasty/client/index.html?q=" + accessionNumberJTextField.getText());
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.DEFAULT_CURSOR));
    }//GEN-LAST:event_loadDastyJButtonActionPerformed

    private void loadUniProtJButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_loadUniProtJButtonActionPerformed
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.WAIT_CURSOR));
        BareBonesBrowserLaunch.openURL("http://www.uniprot.org/uniprot/" + accessionNumberJTextField.getText());
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.DEFAULT_CURSOR)); 
    }//GEN-LAST:event_loadUniProtJButtonActionPerformed

    private void loadReactomeJButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_loadReactomeJButtonActionPerformed
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.WAIT_CURSOR));
        BareBonesBrowserLaunch.openURL("http://www.reactome.org/cgi-bin/link?SOURCE=UNIPROT&ID=" + accessionNumberJTextField.getText());
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.DEFAULT_CURSOR)); 
    }//GEN-LAST:event_loadReactomeJButtonActionPerformed

    private void loadStringJButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_loadStringJButtonActionPerformed
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.WAIT_CURSOR));
        BareBonesBrowserLaunch.openURL("http://string-db.org/newstring_cgi/show_input_page.pl?identifier=" + accessionNumberJTextField.getText());
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.DEFAULT_CURSOR));
    }//GEN-LAST:event_loadStringJButtonActionPerformed

    private void loadDavidJButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_loadDavidJButtonActionPerformed
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.WAIT_CURSOR));
        BareBonesBrowserLaunch.openURL("http://david.abcc.ncifcrf.gov/api.jsp?type=UNIPROT_ACCESSION&ids=" + accessionNumberJTextField.getText() + "&tool=summary");
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.DEFAULT_CURSOR));
    }//GEN-LAST:event_loadDavidJButtonActionPerformed

    private void loadDavidJButton1ActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_loadDavidJButton1ActionPerformed
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.WAIT_CURSOR));
        BareBonesBrowserLaunch.openURL("http://www.ebi.ac.uk/intact/pages/interactions/interactions.xhtml?query=" + accessionNumberJTextField.getText());
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.DEFAULT_CURSOR));
    }//GEN-LAST:event_loadDavidJButton1ActionPerformed

    private void loadDavidJButton2ActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_loadDavidJButton2ActionPerformed
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.WAIT_CURSOR));
        BareBonesBrowserLaunch.openURL("http://www.ebi.ac.uk/QuickGO/GSearch?q=" + accessionNumberJTextField.getText() + "&what=Protein");
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.DEFAULT_CURSOR));
    }//GEN-LAST:event_loadDavidJButton2ActionPerformed

    private void loadDavidJButton3ActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_loadDavidJButton3ActionPerformed
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.WAIT_CURSOR));
        BareBonesBrowserLaunch.openURL("http://www.ebi.ac.uk/interpro/ISearch?query=" + accessionNumberJTextField.getText());
        this.setCursor(new java.awt.Cursor(java.awt.Cursor.DEFAULT_CURSOR));
    }//GEN-LAST:event_loadDavidJButton3ActionPerformed

    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JTextField accessionNumberJTextField;
    private javax.swing.JTextField accessionNumberJTextField1;
    private javax.swing.JTextField accessionNumberJTextField2;
    private javax.swing.JTextField accessionNumberJTextField5;
    private javax.swing.JPanel annotationLinksJPanel;
    private javax.swing.JPanel annotationLinksJPanel1;
    private javax.swing.JPanel annotationLinksJPanel2;
    private javax.swing.JPanel annotationLinksJPanel3;
    private javax.swing.JPanel annotationLinksJPanel4;
    private javax.swing.JPanel annotationLinksJPanel5;
    private javax.swing.JPanel annotationLinksJPanel6;
    private javax.swing.JPanel annotationLinksJPanel7;
    private javax.swing.JPanel annotationLinksJPanel8;
    private javax.swing.JLabel jLabel1;
    private javax.swing.JLabel jLabel10;
    private javax.swing.JLabel jLabel11;
    private javax.swing.JLabel jLabel12;
    private javax.swing.JLabel jLabel13;
    private javax.swing.JLabel jLabel14;
    private javax.swing.JLabel jLabel15;
    private javax.swing.JLabel jLabel17;
    private javax.swing.JLabel jLabel2;
    private javax.swing.JLabel jLabel3;
    private javax.swing.JLabel jLabel4;
    private javax.swing.JLabel jLabel5;
    private javax.swing.JLabel jLabel7;
    private javax.swing.JLabel jLabel8;
    private javax.swing.JLabel jLabel9;
    private javax.swing.JPanel jPanel1;
    private javax.swing.JScrollPane jScrollPane1;
    private javax.swing.JTextArea jTextArea1;
    private javax.swing.JButton loadDastyJButton;
    private javax.swing.JButton loadDavidJButton;
    private javax.swing.JButton loadDavidJButton1;
    private javax.swing.JButton loadDavidJButton2;
    private javax.swing.JButton loadDavidJButton3;
    private javax.swing.JButton loadReactomeJButton;
    private javax.swing.JButton loadStringJButton;
    private javax.swing.JButton loadUniProtJButton;
    // End of variables declaration//GEN-END:variables

    public void setAccessionNumber(String accessionNumber) {
        accessionNumberJTextField.setText(accessionNumber);
    }
}
