#summary Short summary of changes for each version of PeptideShaker.
#labels Featured

----

*Changes in PeptideShaker 0.11.1 (November 21. 2011):*
 * FEATURE IMPROVEMENT: Reduced the time required to save a project using the `Save` option compared to the `Save As` option.
 * FEATURE IMPROVEMENT: "N/A" is now shown in the retention time columns if the retention time is -1.
 * FEATURE IMPROVEMENT: Background highlighting of modifications in the sequence fragmentation plots with tooltips, similar to what us used in the tables.
 * FEATURE IMPROVEMENT: The spectrum sliders (accuracy and intensity level) are now hidden by default.
 * FEATURE IMPROVEMENT: The p-value shown in the GO Enrichment tab is now the FDR-corrected p-value.
 * BUG FIX: Fixed some bugs related to changing the protein inference in the protein table in the Overview tab.
 * BUG FIX: Made sure that the color coding for protein inference is the same in both the Overview tab and the 3D Structures tab.
 * BUG FIX: Fixed a bug with the sorting of the Log2 Diff column in the GO Enrichemnt tab, that sorted on absolute value.
 * LIBRARY UPDATE: Updated jsparklines to version 0.5.27.
 * LIBRARY UPDATE: Updated utilities to version 3.3.3.

----

*Changes in PeptideShaker 0.11.0 (November 17. 2011):*

 * NEW FEATURE: Added a GO Enrichment Analysis tab.
 * NEW FEATURE: Added molecular weight to the protein table in the Overview tab.
 * NEW FEATURE: Added Save in addition to Save As in the File menu.
 * NEW FEATURE: Added Tip of the Day to the Waiting dialog.
 * NEW FEATURE: The modified residues are now indicated by a star above the given residue in the sequence fragmentation plot. 
 * NEW FEATURE: Added the option of hiding proteins in the Overview tab. 
 * NEW FEATURE: Added the option of highlighting/marking proteins in the Overview tab.
 * NEW FEATURE: Made the recent projects appear in the welcome dialog. Also updated the Welcome dialog GUI.
 * NEW FEATURE: The main tabs are now loaded individually, i.e., not all at once when opening a project.

 * FEATURE IMPROVEMENT: Made the importing of PeptideShaker projects a bit smarter.
 * FEATURE IMPROVEMENT: Simplified the default PTM short names for oxidation (mox > ox) and phosphorylation (phopho > p), as the long names took up too much space. 
 * FEATURE IMPROVEMENT: The delta masses used when manually de novo sequencing the spectra are now based on the currently selected modifications and not hardcoded to the default set.
 * FEATURE IMPROVEMENT: Replaced HelpWindow with HelpDialog, identical in every way but a JDialog instead of a JFrame, which makes the modal exclusion handling simpler.
 * FEATURE IMPROVEMENT: Made it possible to interact with the tool when the ProteinFilter dialog is being displayed, e.g., select proteins in the Overview tab without having to close the filtering dialog.
 * FEATURE IMPROVEMENT: Fixed some minor GUI details for when using the backup look and feel.
 * FEATURE IMPROVEMENT: The export folder when zipping a PeptideShaker project can now be chosen by the user.
 * FEATURE IMPROVEMENT: Sectrumpanel now displays all spectra.
 * FEATURE IMPROVEMENT: Added horizontal scroll bars to the peptide tables in the PTM tab, displayed when required. 
 * FEATURE IMPROVEMENT: Improved the focus traversal order in the protein filtering dialog.
 * FEATURE IMPROVEMENT: Removed the index column in the ptm selection panel in the Modifications tab to make more room for the peptides tables.
 * FEATURE IMPROVEMENT: Spectrum IDs tab: added more space for the label in the retention time cell renderer to show higher retention time values.
 * FEATURE IMPROVEMENT: Retention time is now shown for the PSMs in the Modification tab.
 * FEATURE IMPROVEMENT: Improved the error handling if a required mgf file is missing when trying open a new project.
 * FEATURE IMPROVEMENT: Moved the identification filter settings in a separate dialog.

 * BUG FIX: Fixed a GUI bug with the spectrum slidersin the Spectrum IDs tab.
 * BUG FIX: Fixed a bug in the Da/ppm selection when importing the value from a SearchGUI properties file (it always defaulted to ppm...). 
 * BUG FIX: Fixed some minor bugs related to the export PeptideShaker project to zip file feature.
 * BUG FIX: Fixed a bug in the ProteinInferencePeptideLevelDialog where one of the tables was editable, and thus the UniProt link didn't work.
 * BUG FIX: Minor GUI fixes for the PSM column in the Spectrum IDs tab.
 * BUG FIX: Fixed a problem with the location of the popup menus in the contextual export options when the tool was not maximized.
 * BUG FIX: Fixed a bug where the spectrum annotations could not be changed from the annotation menu bar in the Spectrum IDs and PTM tabs.
 * BUG FIX: Fixed a bug where the tool was not always reset correctly when opening a new project.
 * BUG FIX: Fixed a problem with the validation tab when editing both threshold and PEP window at the same time and then moving to a different tab.

 * LIBRARY UPDATE: Updated the UniProt API to version 2011.11.
 * LIBRARY UPDATE: Updated utilities to version 3.3.2.
 * LIBRARY UPDATE: Updated jsparklines to version 0.5.26.

Download Count: 7

----

*Changes in PeptideShaker 0.10.3 (October 24. 2011):*
 * BUG FIX: The help text in the Modifications tab is now visible.
 * LIBRARY UPDATE: Updated utilities to version 3.2.22, which now also supports the special amino acid Selenocysteine (U).

Download Count: 30

----

*Changes in PeptideShaker 0.10.2 (October 21. 2011):*
 * BUG FIX: Fixed a minor GUI issue related to the left hand part of the spectrum panel not aligning correctly when the peptide/psm panel was hidden.
 * BUG FIX: Fixed some resizing issues for modification profile plots in the modification tab where part of the profile could be cut off if one profile was longer than the other.
 * BUG FIX: Fixed some resizing issues in the Spectrum ID and Validation tabs.
 * LIBRARY UPDATE: Updated utilities to version 3.2.21, which now supports the special Mascot amino acids B, Z and X.

Download Count: 4

----

*Changes in PeptideShaker 0.10.1 (October 20. 2011):*
 * NEW FEATURE: Added hide options to all the contextual menus in the Overview tab.
 * LIBRARY UPDATE: Updated uniprotjapi to version 2011.10, fixing a problem with the UniProt connection.

Download Count: 5

----

*Changes in PeptideShaker 0.10.0 (October 19. 2011):*
 * NEW FEATURE: QC plot tab added.
 * NEW FEATURE: Annotation tab with links to other protein resources.
 * NEW FEATURE: Made it possible to use relative of absolute error in the mass error and bubble plots.
 * NEW FEATURE: Added a new sequence coverage plot with all the found peptides indicated.
 * NEW FEATURE: PTM location confidence scoring.
 * NEW FEATURE: Added a Welcome dialog.
 * NEW FEATURE: Protein inference is now displayed also at the peptide level.
 * NEW FEATURE: PTM color coding.
 * NEW FEATURE: Modification profile plots.
 * NEW FEATURE: The 10 last recently opened projects can now be opened directly from the File menu.
 * NEW FEATURE: Simpe export of identification features, and contextual exports for each panel.
 * NEW FEATURE: FASTA and spectrum files are now indexed and not loaded into memory anymore.
 * NEW FEATURE: Identification matches are stored in a separate folder and not loaded in the memory anymore.
 * NEW FEATURE: Annotation intensity level slider and annotation accuracy sliders added to the spectrum plots.
 * FEATURE IMPROVEMENT: A global maximum is now used for the sparklines for all charge and mass error columns.
 * FEATURE IMPROVEMENT: Updated most of the help texts.
 * FEATURE IMPROVEMENT: The peptides are now sorted according to their number of spectra.
 * FEATURE IMPROVEMENT: Modifications are now shown directly on the peptide sequence. Modifications are now also mapped onto the 3D structure.
 * FEATURE IMPROVEMENT: Y-axis auto zoom in the spectrum plots now disregards unannotated peaks (unless all peaks are shown).
 * FEATURE IMPROVEMENT: Changed the default -Mxm option from 1500M to 1024M, to make sure that the tool starts on all machines.
 * FEATURE IMPROVEMENT: Spectrum annotation selection has been made simpler by the addition of an annotation menu bar.
 * FEATURE IMPROVEMENT: Added a secondary progress bar in the WaitingDialog showing progress for the current process.
 * FEATURE IMPROVEMENT: Added terminals to the sequence fragmentation plots.
 * FEATURE IMPROVEMENT: Use of subscript for the numbers in all peak annotations in all plots.
 * FEATURE IMPROVEMENT: Improved automatic spectrum annotation.
 * FEATURE IMPROVEMENT: The mgf paths in the search result files OS independent.
 * FEATURE IMPROVEMENT: The two residues in front of and after the selected peptide is now shown in the panel title of the spectrum panel.
 * FEATURE IMPROVEMENT: Improved the way exceptions related to proteins not found in the FASTA file are handled. PeptideShaker now stops the loading of a project if such an exception is thrown.
 * BUG FIX: Removed peptides without PSMs from the display.
 * LIBRARY UPDATE: Updated utilities to 3.2.20.
 * LIBRARY UPDATE: Updated jsparklines to version 0.5.20.
 * LIBRARY UPDATE: Updated xtandem parser to version 1.3.5. 
 * LIBRARY UPDATE: Updated xtandem parser to version 1.4.6. 
 * LIBRARY UPDATE: Updated mascotdatfile to version 3.2.6.

Download Count: 5

----

*Changes in PeptideShaker 0.9.3 (July 17. 2011):*
 * FEATURE IMPROVEMENT: The link between the Overview and the Structure tab is now smarter, and updates less frequently.
 * FEATURE IMPROVEMENT: The maximum initial Java memory size is now set to 1500M (the magic number for 32 bit Java...).
 * FEATURE IMPROVEMENT: Improved the wrapper so that it now defaults to using the 64 bit Java version if available.
 * BUG FIX: Fixed a bug in the dialog displayed when detecting an unknown protein, where the title and the message was interchanged.
 * BUG FIX: The files selection dialog (for multiple SearchGUI property files) is now located relative to its parent.

Download Count: 121

----

*Changes in PeptideShaker 0.9.2 (July 14. 2011):*
 * BUG FIX: Corrected a minor bug in the preferences dialog. 
 * LIBRARY UPDATE: Updated utilities to 3.1.30, for more compatible databases.

Download Count: 10

----

*Changes in PeptideShaker 0.9.1 (July 14. 2011):*
 * NEW FEATURE: Added protein HTML links to all columns displaying protein accession numbers.
 * NEW FEATURE: Made it possible to include more than one protein HTML link in the same cell in the tables.
 * NEW FEATURE: The 3D protein model now rotates slowly as a default.
 * FEATURE IMPROVEMENT: Minor GUI updates to the protein inferences dialog.
 * FEATURE IMPROVEMENT: Changed all the protein links from pointing the the SRS web page to pointing to UniProt.
 * FEATURE IMPROVEMENT: Extended the 3D Structure help text.
 * BUG FIX: Fixed various minor GUI issues related to using the backup look and feel, i.e., not using Nimbus.
 * LIBRARY UPDATE: Updated utilities to 3.1.29, fixing a bug in the XTandem parsing.

Download Count: 1

----

*Changes in PeptideShaker 0.9 (July 11. 2011):*
 * The first public beta release of PeptideShaker.

Download Count: 6

----